### PR TITLE
fix(externalsecret): infinite reconcile loop with Merge secret

### LINKED ISSUE #2033: ExternalSecret keeps updating with CreationPolicy on Merge
Hello,

we are using External Secrets with Vault and our problem is that External Secrets is spamming our Vault logs with access. I could track the excessive access back to only some of our `ExternalSecrets`. It looks like that when we use `ExternalSecrets` with the `CreationPolicy` on `Merge` to add Vault secrets to existing `Secrets` the `ExternalSecret` keeps updating excessively. 

I could view this by using: `kubectl describe ExternalSecret my-external-secret`
The event section shows a very high amount of Updates for the described `ExternalSecrets`, for example:
image

Since we are using the default `refreshTime` of one hour it should normally be a lot lower. For example the next Screenshot is the Event Section of an `ExternalSecret` which creates a new `Secret` in Kubernetes with the `CreationPolicy: owner`:
image

You can reproduce this issue in your cluster with following yaml:
```
apiVersion: v1
kind: Secret
metadata:
  name: test-secret
  namespace: testing
type: Opaque
stringData:
  test: testing
---
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: test-external-secret
  namespace: testing
spec:
  secretStoreRef:
    name: your-vault-secret-store
    kind: ClusterSecretStore
  target:
    name: test-secret
    creationPolicy: 'Merge'
  data:
    - secretKey: test.key
      remoteRef:
        key: path/to/your/key
        property: key
```

I tested it with the External Secret version of 0.7.2 and Kubernetes version of 1.24.8. Our Kubernetes is running on-premise and is deployed with RKE2.

### PR DESCRIPTION
**Pull request by agaudreault-jive**
## Problem Statement

When `creationPolicy: Merge` is used and the data is merged in a secret with existing data, the generated value for the data-hash annotation is invalid, causing an infinite loop of reconciliation for this secret

**NOTE**: I was not able to find how to run `pkg/controllers/externalsecret/suite_test.go` 

## Related Issue

Fixes #2033

## Proposed Changes

Generate the data hash based on the whole data and not only the keys to patch.

I tested manually in a kind cluster with the following resource. 
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: test-secret
type: Opaque
stringData:
  test: testing
---
apiVersion: external-secrets.io/v1beta1
kind: SecretStore
metadata:
  name: "test"
spec:
  provider:
    fake:
      data:
        - key: /foo/bar
          valueMap:
            foo: bar
---
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: test-external-secret
  annotations:
    test: "1"
spec:
  secretStoreRef:
    name: test
    kind: SecretStore
  target:
    name: test-secret
    creationPolicy: "Merge"
    template:
      metadata:
        annotations: {}
  dataFrom:
    - extract:
        key: "/foo/bar"

```

I added some logs while debuging and before the change, you can see the annotation was always updated to `5211b9bd4ac724a48a80e741f54ca581`, but the annotation was expected to be `2e668630947b1947efc78b05339dcae8`
```
{"level":"info","ts":1689955216.9808261,"logger":"controllers.ExternalSecret","msg":"Get annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"2e668630947b1947efc78b05339dcae8"}
{"level":"info","ts":1689955216.987351,"logger":"controllers.ExternalSecret","msg":"Update annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"5211b9bd4ac724a48a80e741f54ca581"}
{"level":"info","ts":1689955217.0070539,"logger":"controllers.ExternalSecret","msg":"Get annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"2e668630947b1947efc78b05339dcae8"}
{"level":"info","ts":1689955217.0095513,"logger":"controllers.ExternalSecret","msg":"Update annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"5211b9bd4ac724a48a80e741f54ca581"}
{"level":"info","ts":1689955217.020322,"logger":"controllers.ExternalSecret","msg":"Get annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"2e668630947b1947efc78b05339dcae8"}
{"level":"info","ts":1689955217.022314,"logger":"controllers.ExternalSecret","msg":"Update annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"5211b9bd4ac724a48a80e741f54ca581"}
```

After the changes you can see that the annotation is updated to the correct value
```
{"level":"info","ts":1689956737.0324805,"logger":"controllers.ExternalSecret","msg":"Update annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"2e668630947b1947efc78b05339dcae8"}
{"level":"info","ts":1689956737.0448456,"logger":"controllers.ExternalSecret","msg":"Get annotation","ExternalSecret":{"name":"test-external-secret","namespace":"default"},"annotation":"2e668630947b1947efc78b05339dcae8"}
```

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: pkg/controllers/externalsecret/externalsecret_controller.go [hunk 1]
```diff
@@ -434,6 +436,10 @@ func hashMeta(m metav1.ObjectMeta) string {
 	})
 }
 
+func hashData(data map[string][]byte) string {
```
Comment (shuheiktgw) [inline pkg/controllers/externalsecret/externalsecret_controller.go:439]: I don't think we need a wrapper for `utils.ObjectHash` should just call `utils.ObjectHash` directly 🙂

File: pkg/controllers/externalsecret/externalsecret_controller.go [hunk 2]
```diff
@@ -541,12 +547,23 @@ func isSecretValid(existingSecret v1.Secret) bool {
 	}
 
 	// if the calculated hash is different from the calculation, then it's invalid
-	if existingSecret.Annotations[esv1beta1.AnnotationDataHash] != utils.ObjectHash(existingSecret.Data) {
+	if existingSecret.Annotations[esv1beta1.AnnotationDataHash] != hashData(existingSecret.Data) {
```
Comment (shuheiktgw) [inline pkg/controllers/externalsecret/externalsecret_controller.go:550]: This does not seem correct to me. In the PR `existingSecret.Annotations[esv1beta1.AnnotationDataHash]` was computed by taking a map of secret data but it compares data directly.

**Comment by shuheiktgw** [comment]
I've been thinking about how we should fix this problem and could not figure out the best way yet... The source of the problem is that we should somehow precompute the hash of the secret data even when we patch it. The current approach is using a map to compute it, but it implicitly assumes a certain way of patching, so I'm not too sure if that is the best or not. Another approach may be applying the patch, getting the result of it, computing the hash, and updating the secret again to attach the hash to it. This requires calling API twice but it is much simpler and robust.

File: pkg/controllers/externalsecret/externalsecret_controller.go (hunk 2 above)
Comment (agaudreault-jive) [reply pkg/controllers/externalsecret/externalsecret_controller.go:550]: The value of the annotation is equivalent to the merge of existing data + new data.
Previously, the annotation value was equal to new data only. 

Then the secret is updated.

On the next iteration, the value of the annotation will be equal to the has of the current data.

Comment (shuheiktgw) [reply pkg/controllers/externalsecret/externalsecret_controller.go:550]: Ah, sorry you are right!

File: pkg/controllers/externalsecret/externalsecret_controller.go [hunk 3]
```diff
@@ -541,12 +547,23 @@ func isSecretValid(existingSecret v1.Secret) bool {
...
 		return false
 	}
 	return true
 }
 
+func (r *Reconciler) setDataHashAnnotation(existing *v1.Secret, secret *v1.Secret) {
```
Comment (shuheiktgw) [inline pkg/controllers/externalsecret/externalsecret_controller.go:556]: nit, Removing side-effects makes it more readable. I would write 

```
func (r *Reconciler) computeDataHashAnnotation(existing *v1.Secret, secret *v1.Secret) string
```

and

```
secret.Annotations[esv1beta1.AnnotationDataHash] = r.computeDataHashAnnotation(&existingSecret, secret)
```

in the mutationFunc.

**Comment by shuheiktgw** [comment]
@agaudreault-jive Seems like multiple CI Jobs failed. Would you take a look and fix them? Also, I believe we need a unit test to verify the new behavior 🙂

**Comment by agaudreault-jive** [comment]
> @agaudreault-jive Seems like multiple CI Jobs failed. Would you take a look and fix them? Also, I believe we need a unit test to verify the new behavior 🙂
> 
> > NOTE: I was not able to find how to run pkg/controllers/externalsecret/suite_test.go
> 
> Running `make test` is the easiest way to run unit tests locally or you can run the following command if you want to run only external secret controller tests.
> 
> ```
> KUBEBUILDER_ASSETS=$(setup-envtest use 1.24.x -p path) go test -v -race ./pkg/controllers/externalsecret/...
> ```
> 
> Please make sure that you installed [setup-envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/tools/setup-envtest) in your environment. Let me know if you need any help!

Here is the result I get locally. Even if I manually add incorrect values in the test for the `pkg/controllers/externalsecret/externalsecret_controller_test.go` file, the test seems to pass so I am unable to trigger them properly. I also see a lot of messages like this during the test output.

```
E0724 09:42:20.084396   87508 event.go:280] Server rejected event '&v1.Event{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:v1.ObjectMeta{Name:"test-es.1774d1877ffc1e30", GenerateName:"", Namespace:"ctrl-test-test-nsjw7r7", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), DeletionTimestamp:<nil>, DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Finalizers:[]string(nil), ManagedFields:[]v1.ManagedFieldsEntry(nil)}, InvolvedObject:v1.ObjectReference{Kind:"ExternalSecret", Namespace:"ctrl-test-test-nsjw7r7", Name:"test-es", UID:"c5c57dfe-3d5a-4aa3-9de3-4a6934a024c2", APIVersion:"external-secrets.io/v1beta1", ResourceVersion:"842", FieldPath:""}, Reason:"UpdateFailed", Message:"secrets \"test-secret\" is forbidden: unable to create new content in namespace ctrl-test-test-nsjw7r7 because it is being terminated", Source:v1.EventSource{Component:"external-secrets", Host:""}, FirstTimestamp:time.Date(2023, time.July, 24, 9, 42, 20, 68798000, time.Local), LastTimestamp:time.Date(2023, time.July, 24, 9, 42, 20, 81992000, time.Local), Count:2, Type:"Warning", EventTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Series:(*v1.EventSeries)(nil), Action:"", Related:(*v1.ObjectReference)(nil), ReportingController:"", ReportingInstance:""}': 'events "test-es.1774d1877ffc1e30" is forbidden: unable to create new content in namespace ctrl-test-test-nsjw7r7 because it is being terminated' (will not retry!)
```

```
Ran 74 of 74 Specs in 65.338 seconds
SUCCESS! -- 74 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestAPIs (65.34s)
=== RUN   TestGetExternalSecretCondition
=== RUN   TestGetExternalSecretCondition/Status_has_a_condition_of_the_specified_type
=== RUN   TestGetExternalSecretCondition/Status_does_not_have_a_condition_of_the_specified_type
--- PASS: TestGetExternalSecretCondition (0.00s)
    --- PASS: TestGetExternalSecretCondition/Status_has_a_condition_of_the_specified_type (0.00s)
    --- PASS: TestGetExternalSecretCondition/Status_does_not_have_a_condition_of_the_specified_type (0.00s)
=== RUN   TestSetExternalSecretCondition
=== RUN   TestSetExternalSecretCondition/ExternalSecret_has_no_conditions
=== RUN   TestSetExternalSecretCondition/ExternalSecret_already_has_the_same_condition
=== RUN   TestSetExternalSecretCondition/ExternalSecret_has_a_different_condition_with_the_same_type_and_status
=== RUN   TestSetExternalSecretCondition/ExternalSecret_has_a_different_condition
--- PASS: TestSetExternalSecretCondition (0.00s)
    --- PASS: TestSetExternalSecretCondition/ExternalSecret_has_no_conditions (0.00s)
    --- PASS: TestSetExternalSecretCondition/ExternalSecret_already_has_the_same_condition (0.00s)
    --- PASS: TestSetExternalSecretCondition/ExternalSecret_has_a_different_condition_with_the_same_type_and_status (0.00s)
    --- PASS: TestSetExternalSecretCondition/ExternalSecret_has_a_different_condition (0.00s)
PASS
ok      github.com/external-secrets/external-secrets/pkg/controllers/externalsecret     66.464s
```

**Comment by shuheiktgw** [comment]
Hi, sorry that it took me a while to get back to you 🙇 

> I also see a lot of messages like this during the test output.

The error messages above are fine and expected. That happens since the controller publishes events in a namespace being deleted! The log you posted suggests `TestAPIs`, which includes the Ginkgo tests passed without any problems, so I believe it is running correctly in your local environment 🤔

**Comment by shuheiktgw** [comment]
Hi @agaudreault-jive 👋 Are you still working on this PR? Please let me know if you need any help to proceed!

**Comment by agaudreault-jive** [comment]
@shuheiktgw I was able to run and update the tests properly.

File: pkg/controllers/externalsecret/externalsecret_controller_test.go [hunk 4]
```diff
@@ -1730,6 +1727,28 @@ var _ = Describe("ExternalSecret controller", Serial, func() {
...
+				Namespace: ExternalSecretNamespace,
+			},
+			Data: map[string][]byte{
+				existingKey: []byte(existingVal),
+			},
+		}, client.FieldOwner(FakeManager))).To(Succeed())
+
+		fakeProvider.WithGetSecret([]byte(secretVal), nil)
+		tc.checkSecret = func(es *esv1beta1.ExternalSecret, secret *v1.Secret) {
+			Expect(secret.Annotations[esv1beta1.AnnotationDataHash]).To(Equal("82a8288e398d5354ab110d2244b7c0b1"))
```
Comment (shuheiktgw) [inline pkg/controllers/externalsecret/externalsecret_controller_test.go:1748]: Using the hash value directly is hard to maintain and understand. Would you create the hash value using `utils.ObjectHash`?

**Review by shuheiktgw — CHANGES_REQUESTED** [review]
@agaudreault-jive Thanks for the great work! I left one comment but overall it looks good to me 🙂

File: pkg/controllers/externalsecret/externalsecret_controller_test.go (hunk 4 above)
Comment (agaudreault-jive) [reply pkg/controllers/externalsecret/externalsecret_controller_test.go:1748]: Sure. I will refactor other tests as well 👍

Comment (agaudreault-jive) [reply pkg/controllers/externalsecret/externalsecret_controller_test.go]: Done!

**Comment by shuheiktgw** [comment]
SonarCloud failure does not seem to relate to this PR.

**Comment by shuheiktgw** [comment]
LGTM! I'm going to ask for a review from other maintainers 🙂

**Comment by agaudreault-jive** [comment]
@shuheiktgw Any updates on this?

**Comment by shuheiktgw** [comment]
Sorry, I've asked for a review again!
https://kubernetes.slack.com/archives/C047LA9MUPJ/p1693009454748039

**Comment by agaudreault-jive** [comment]
@moolen any way to cut a release with this? :)

**Comment by moolen** [comment]
Sure, i'll do. But i want 1-2 other PRs released as well. Probably end of this week / beginning of next week.

**Pull request merged**